File: it-agents/skills/code-architect/references/code-examples/python/composite.py
"""Composite — compose objects into trees, treat leaf and container uniformly."""

from __future__ import annotations

from abc import ABC, abstractmethod


class Node(ABC):
    @abstractmethod
    def size(self) -> int:
        ...

    @abstractmethod
    def render(self, indent: int = 0) -> str:
        ...


class FileLeaf(Node):
    def __init__(self, name: str, size: int) -> None:
        self.name = name
        self._size = size

    def size(self) -> int:
        return self._size

    def render(self, indent: int = 0) -> str:
        return f"{'  ' * indent}- {self.name} ({self._size})"


class FolderComposite(Node):
    def __init__(self, name: str) -> None:
        self.name = name
        self._children: list[Node] = []

    def add(self, node: Node) -> "FolderComposite":
        self._children.append(node)
        return self

    def size(self) -> int:
        return sum(child.size() for child in self._children)

    def render(self, indent: int = 0) -> str:
        lines = [f"{'  ' * indent}+ {self.name}/ ({self.size()})"]
        lines += [child.render(indent + 1) for child in self._children]
        return "\n".join(lines)


if __name__ == "__main__":
    root = FolderComposite("root")
    root.add(FileLeaf("a.txt", 10))
    sub = FolderComposite("sub").add(FileLeaf("b.txt", 5)).add(FileLeaf("c.txt", 7))
    root.add(sub)
    print(root.render())

File: it-agents/skills/code-architect/references/code-examples/cpp/Composite.cpp
// Composite — compose objects into trees, treat leaf and container uniformly.
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class Node {
public:
    virtual ~Node() = default;
    virtual int weight() const = 0;
    virtual void print(int depth = 0) const = 0;
protected:
    static void indent(int depth) {
        for (int i = 0; i < depth; ++i) std::cout << "  ";
    }
};

// Leaf.
class Item : public Node {
    std::string name;
    int w;
public:
    Item(std::string n, int weight) : name(std::move(n)), w(weight) {}
    int weight() const override { return w; }
    void print(int depth = 0) const override {
        indent(depth);
        std::cout << "- " << name << " (" << w << ")\n";
    }
};

// Composite.
class Group : public Node {
    std::string name;
    std::vector<std::shared_ptr<Node>> children;
public:
    explicit Group(std::string n) : name(std::move(n)) {}
    void add(std::shared_ptr<Node> child) { children.push_back(std::move(child)); }
    int weight() const override {
        int total = 0;
        for (const auto& c : children) total += c->weight();
        return total;
    }
    void print(int depth = 0) const override {
        indent(depth);
        std::cout << "+ " << name << " (" << weight() << ")\n";
        for (const auto& c : children) c->print(depth + 1);
    }
};

int main() {
    auto root = std::make_shared<Group>("root");
    auto sub = std::make_shared<Group>("sub");
    sub->add(std::make_shared<Item>("a", 3));
    sub->add(std::make_shared<Item>("b", 4));
    root->add(sub);
    root->add(std::make_shared<Item>("c", 5));

    root->print();
    std::cout << "total weight = " << root->weight() << "\n";
    return 0;
}
